addons/stock/models/stock_move.py

```diff
@@ -0,0 +1,1109 @@
+                            qty = min(lot_qty[lot], move_qty)
+                            quants = quant_obj.quants_get_preferred_domain(cr, uid, qty, move, ops=ops, lot_id=lot, domain=domain, preferred_domain_list=[], context=context)
+                            quant_obj.quants_reserve(cr, uid, quants, move, record, context=context)
+                            lot_qty[lot] -= qty
+                            move_qty -= qty
+
+        for move in todo_moves:
+            if move.linked_move_operation_ids:
+                continue
+            #then if the move isn't totally assigned, try to find quants without any specific domain
+            if move.state != 'assigned':
+                qty_already_assigned = move.reserved_availability
+                qty = move.product_qty - qty_already_assigned
+                quants = quant_obj.quants_get_preferred_domain(cr, uid, qty, move, domain=main_domain[move.id], preferred_domain_list=[], context=context)
+                quant_obj.quants_reserve(cr, uid, quants, move, context=context)
+
+        #force assignation of consumable products and incoming from supplier/inventory/production
+        # Do not take force_assign as it would create pack operations
+        if to_assign_moves:
+            self.write(cr, uid, list(to_assign_moves), {'state': 'assigned'}, context=context)
+        if not no_prepare:
+            self.check_recompute_pack_op(cr, uid, ids, context=context)
+
+    @api.multi
+    def action_cancel(self):
+        """ Cancels the moves and if all moves are cancelled it cancels the picking. """
+        # TDE DUMB: why is cancel_procuremetn in ctx we do quite nothing ?? like not updating the move ??
+        if any(move.state == 'done' for move in self):
+            raise UserError(_('You cannot cancel a stock move that has been set to \'Done\'.'))
+
+        procurements = self.env['procurement.order']
+        for move in self:
+            if move.reserved_quant_ids:
+                move.quants_unreserve()
+            if self.env.context.get('cancel_procurement'):
+                if move.propagate:
+                    procurements.search([('move_dest_id', '=', move.id)]).cancel()
+            else:
+                if move.move_dest_id:
+                    if move.propagate:
+                        move.move_dest_id.action_cancel()
+                    elif move.move_dest_id.state == 'waiting':
+                        # If waiting, the chain will be broken and we are not sure if we can still wait for it (=> could take from stock instead)
+                        move.move_dest_id.write({'state': 'confirmed'})
+                if move.procurement_id:
+                    procurements |= move.procurement_id
+
+        self.write({'state': 'cancel', 'move_dest_id': False})
+        if procurements:
+            procurements.check()
+        return True
+
+    @api.cr_uid_ids_context
+    def recalculate_move_state(self, cr, uid, move_ids, context=None):
```

Comment:
Can we change method signature to pure @api-multi (i.e, only **self** as an argument)?